File: backend/modules/suppliers/application/remove_product_from_supplier_use_case.py
from modules.suppliers.domain.interfaces.repositories.i_supplier_repository import (
    ISupplierRepository,
)
from modules.suppliers.domain.interfaces.use_cases.i_remove_product_from_supplier_use_case import (
    IRemoveProductFromSupplierUseCase,
)


class RemoveProductFromSupplierUseCase(IRemoveProductFromSupplierUseCase):
    def __init__(self, repo: ISupplierRepository) -> None:
        self._repo = repo

    async def execute(self, supplier_id: int, product_id: int) -> None:
        await self._repo.remove_product(supplier_id, product_id)

File: frontend/src/app/domain/usecases/supplier-product/remove-product-from-supplier.usecase.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { SupplierProductRepository } from '@domain/repositories/supplier-product.repository';
import { SupplierProductValidationError } from '@domain/models/supplier-product-errors';

@Injectable({ providedIn: 'root' })
export class RemoveProductFromSupplierUseCase {
  private readonly supplierProductRepository = inject(SupplierProductRepository);

  execute(supplierId: number, productId: number): Observable<void> {
    if (!Number.isInteger(supplierId) || supplierId <= 0) {
      throw new SupplierProductValidationError({ supplierId }, 'Invalid supplier ID.');
    }

    if (!Number.isInteger(productId) || productId <= 0) {
      throw new SupplierProductValidationError({ productId }, 'Invalid product ID.');
    }

    return this.supplierProductRepository.removeProductFromSupplier(supplierId, productId);
  }
}
